File: algorithms/java/解码方法/解码方法.java
// Source : https://leetcode.cn/problems/decode-ways/
// Author : Yanan Wang
// Date   : 2023-06-14

class Solution {
    public int numDecodings(String s) {
        int n = s.length();
        int[] dp = new int[n + 1];
        dp[0] = 1;
        for (int i = 1; i <= n; i++) {
            if (check(s.charAt(i - 1))) {
                dp[i] += dp[i - 1];
            }
            if (i > 1 && check(s.charAt(i - 2), s.charAt(i - 1))) {
                dp[i] += dp[i - 2];
            }
        }
        return dp[n];
    }
    
    public boolean check(char ch) {
        return ch == '0' ? false : true;
    }

    public boolean check(char ch1, char ch2) {
        return (ch1 == '1' || (ch1 == '2' && ch2 <= '6')) ? true : false;
    }
}

File: algorithms/python/解码方法/解码方法.py
# Source : https://leetcode.cn/problems/decode-ways/description/ 
# Author : Yanan Wang
# Date   : 2023-10-30

class Solution:
    def numDecodings(self, s: str) -> int:
        n = len(s)
        dp = [1] + [0] * n
        for i in range(1, n + 1):
            if s[i - 1] != '0':
                dp[i] = dp[i - 1]
            if i > 1 and s[i - 2] != '0' and int(s[i-2 : i]) <= 26:
                dp[i] += dp[i - 2]
        return dp[n] 
